datajunction-server/datajunction_server/models/access.py

```diff
@@ -0,0 +1,437 @@
+        """
+        self.add_request(
+            ResourceRequest(
+                verb=verb or self.base_verb,
+                access_object=ResourceObjectBase.from_namespace(namespace),
+            ),
+        )
+
+    def raise_if_invalid_requests(self, show_denials: bool = True):
+        """
+        Raises if validate has ever given any invalid requests
+        """
+        denied = ", ".join(
+            [
+                str(request)
+                for request in self.validation_results
+                if not request.approved
+            ],
+        )
+        if denied:
+            message = (
+                f"Authorization of User `{self.user.username if self.user else 'no user'}` "
+                "for this request failed."
+                f"\nThe following requests were denied:\n{denied}."
+                if show_denials
+                else ""
+            )
+            raise DJException(
+                http_status_code=HTTPStatus.FORBIDDEN,
+                errors=[
+                    DJError(
+                        code=ErrorCode.UNAUTHORIZED_ACCESS,
+                        message=message,
+                    ),
+                ],
+            )
+
+    def validate(self) -> Set[ResourceRequest]:
+        """
+        Checks with ACS and stores any returned invalid requests
+        """
+        self.validation_request_count += 1
+
+        access_control = AccessControl(
+            user=self.user.username if self.user is not None else "",
+            state=self.state,
+            direct_requests=deepcopy(self.direct_requests),
+            indirect_requests=deepcopy(self.indirect_requests),
+            validation_request_count=self.validation_request_count,
+        )
+
+        self.validate_access(access_control)  # type: ignore
+
+        self.validation_results = access_control.requests
+
+        if any((result.approved is None for result in self.validation_results)):
+            raise DJException(
+                http_status_code=HTTPStatus.FORBIDDEN,
+                errors=[
+                    DJError(
+                        code=ErrorCode.INCOMPLETE_AUTHORIZATION,
+                        message="Injected `validate_access` must approve or deny all requests.",
+                    ),
+                ],
+            )
+
+        return self.validation_results
+
+    def validate_and_raise(self):
+        """
+        Validates with ACS and raises if any resources were denied
+        """
+        self.validate()
+        self.raise_if_invalid_requests()
+
+
+def validate_access_nodes(
+    validate_access: "ValidateAccessFn",  # pylint: disable=W0621
+    verb: ResourceRequestVerb,
+    user: Optional[User],
+    nodes: Iterable[Union[NodeRevision, Node]],
+    raise_: bool = False,
+) -> List[Union[NodeRevision, Node]]:
+    """
+    Validate the access of the user to a set of nodes
+    """
+    if user is None:
+        return list(nodes)  # pragma: no cover
```

Comment:
More reasons to make auth mandatory 😄 . We'll have to remember to come back and change this back to raising an exception, or remove the `if user is None` check entirely.

Reply:
I intend to remove all these optional checks including any for access_control once the user is required and access control is implemented for all routes